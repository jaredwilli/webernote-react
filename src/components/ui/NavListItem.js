import React from 'react';
import { Link } from 'react-router-dom';

import { shorten } from '../../common/helpers.js';
import { PLURALS } from '../../constants/colors';

function NavListItem({ type, notes = [], items = [], ...props }) {
    // Set the initial count value and normalize the name
    items = items.map(item => {
        item.name = (item.label) ? item.label : item.name;
        return item;
    });

    // Unpluralize the type for notebooks and labels
    type = (type !== 'tags') ? PLURALS[type] : type;

    const notesWithType = (type) => notes.filter(note => note[type] )

    // Set the items with the total count of each item from notesWithType
    const itemWithCount = (item) => notesWithType.reduce((sum, note) => {
        return (note[type].id === item.id) ? sum + 1 : sum;
    }, 0);

    const tagsWithCount = (item) => notesWithType.map(note => {
        let count = note[type].reduce((sum, tag) => {
            return (tag.id === item.id) ? sum + 1 : sum;
        }, 0);
        return count;
    });

    return (
        <ul className={`${type}-list`}>
            {items.map(item =>
                <li key={item.id} id={item.id}>
                    {(type === 'label') &&
                        <div className="note-label" style={{ background: item.hex }} />
                    }
                    <Link to={`/${type}/${item.name.toLowerCase()}`}>
                        <span className="name">{shorten(item.name, 25)}</span>
                    </Link>&nbsp;
                    <span className="count">
                        {(type === 'tags') ? tagsWithCount(item) : itemWithCount(item)}
                    </span>
                </li>
            )}
        </ul>
    );
}

export default NavListItem;
