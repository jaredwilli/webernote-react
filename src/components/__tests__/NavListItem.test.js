import React from 'react';
import renderer from 'react-test-renderer';
import TestUtils from 'react-dom/test-utils'; // ES6

import NavListItem from '../ui/NavListItem.js';

describe('<NavListItem />', () => {
    const type = 'notebooks';
    const items = [];
    const notes = [];

    it('should match the snapshot', () => {
        const component = renderer.create(
            <NavListItem type={type} items={items} notes={notes} />
        ).toJSON();

        expect(component).toMatchSnapshot();


    });
});
