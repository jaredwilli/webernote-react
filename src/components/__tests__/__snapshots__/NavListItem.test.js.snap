// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NavListItem /> should match the snapshot 1`] = `
<ul
  className="notebook-list"
/>
`;
